File: solutions/csharp/perfect-numbers/1/PerfectNumbers.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum Classification
{
    Perfect,
    Abundant,
    Deficient
}

public static class PerfectNumbers
{
    public static int AliquotSum(int number)
    {
        var factors = new List<int>();
        for (int i = 1; i < number; i++)
        {
            if (number % i == 0)
            {
                factors.Add(i);
            }
        }
        return factors.Sum();
    }
    
    public static Classification Classify(int number)
    {
        if (number <= 0)
        {
            throw new ArgumentOutOfRangeException("Only positive integers can be classified.");
        }
        if (AliquotSum(number) == number)
        {
            return Classification.Perfect;
        }
        else if (AliquotSum(number) > number)
        {
            return Classification.Abundant;
        }
        else
        {
            return Classification.Deficient;
        }
    }
}

File: solutions/fsharp/perfect-numbers/1/PerfectNumbers.fs
module PerfectNumbers

type Classification = Perfect | Abundant | Deficient 

let classify n : Classification option =
    if n < 1 then
        None
    elif n = 1 then
        Some Classification.Deficient
    else
        let mutable factors = []
        for i in [1..int(sqrt(float n))] do
            if n % i = 0 then
                factors <- List.append factors [i]
                if n / i <> i && i <> 1 then
                    factors <- List.append factors [n / i]
        let aliquotSum = List.sum factors
        if aliquotSum = n then
            Some Classification.Perfect
        elif aliquotSum > n then
            Some Classification.Abundant
        else
            Some Classification.Deficient
